### PR TITLE
[core] Introduce `ext_dispatch_call` and use node-runtime types

### PR DESCRIPTION
**Pull request by ascjones**
Still outstanding: 

- [ ] Implement `ext_dispatch_call` for `TestEnv`
- [ ] Test contract with dispatch call 
- [ ] Convert other types (`Balance`, `Hash`) to use `node-runtime` types?

### CODE REVIEW

**Comment by Robbepop** [comment]
We probably also want a code review here from @pepyakin .

File: core/Cargo.toml [hunk 1]
```diff
@@ -19,11 +19,15 @@ include = ["/Cargo.toml", "src/**/*.rs", "/README.md", "/LICENSE"]
 
 [dependencies]
 pdsl_utils = { path = "../utils/" }
-parity-codec = { version = "3.2", default-features = false, features = ["derive", "full"] }
+parity-codec = { version = "3.3", default-features = false, features = ["derive", "full"] }
 log = "0.4"
 wee_alloc = { version = "0.4", default-features = false }
+node-runtime = { git = "https://github.com/paritytech/substrate/", branch = "v1.0", package = "node-runtime", default-features = false }
```
Comment (Robbepop) [inline core/Cargo.toml]: So having `node-runtime` in our dependency tree means that we now have a shared dependency base with it which is huge I assume. Since `cargo` cannot properly unshare crate features of dependencies we have to be really careful about overlapping ones.

File: core/src/env/api.rs [hunk 2]
```diff
@@ -39,6 +40,10 @@ pub fn caller() -> Address {
     ContractEnv::caller()
 }
 
+pub fn dispatch_call(value: node_runtime::Call) {
```
Comment (Robbepop) [inline core/src/env/api.rs]: Missing documentation

File: core/src/env/srml/srml_only/impls.rs [hunk 3]
```diff
@@ -94,7 +95,7 @@ where
 impl<T> Env for SrmlEnv<T>
 where
     T: EnvTypes,
-    <T as EnvTypes>::Address: for<'a> TryFrom<&'a [u8]>,
+    <T as EnvTypes>::Address: Decode,
     <T as EnvTypes>::Hash: for<'a> TryFrom<&'a [u8]>,
```
Comment (Robbepop) [inline core/src/env/srml/srml_only/impls.rs]: Those two lines look asymmetric. Would have suspected that `Hash` requires the same interface now. Also I am not sure if we are still in need of `EnvTypes` due to `node-runtime`.

File: core/src/env/srml/types.rs [hunk 4]
```diff
@@ -34,24 +35,8 @@ impl EnvTypes for DefaultSrmlTypes {
...
-    fn try_from(bytes: &'a [u8]) -> Result<Address, TryFromSliceError> {
-        let address = <[u8; 32]>::try_from(bytes)?;
-        Ok(Address(address))
-    }
-}
+/// The default SRML address type
+pub type Address = node_runtime::Address;
 
 /// The default SRML balance type.
 pub type Balance = u64;
```
Comment (Robbepop) [inline core/src/env/srml/types.rs]: Can you convert all pDSL types to `node-runtime` types? So also `Hash` and `Balance` etc.?

File: core/src/env/traits.rs [hunk 5]
```diff
@@ -62,6 +62,8 @@ pub trait Env: EnvTypes + EnvStorage {
     /// Returns the chain address of the caller.
     fn caller() -> <Self as EnvTypes>::Address;
 
+    fn dispatch_call(data: &[u8]);
```
Comment (Robbepop) [inline core/src/env/traits.rs]: Missing documentation

File: core/src/lib.rs [hunk 6]
```diff
@@ -24,18 +24,15 @@
...
-#[global_allocator]
-static ALLOC: wee_alloc::WeeAlloc = wee_alloc::WeeAlloc::INIT;
-
 #[cfg(all(test, feature = "std"))]
 #[macro_use]
 mod test_utils;
 
-#[cfg(not(feature = "std"))]
-mod panic_handler;
-
```
Comment (Robbepop) [inline core/src/lib.rs]: So I guess this functionality is now handled by `node-runtime`. Is there a way to control it from pDSL?

File: core/src/lib.rs [hunk 7]
```diff
@@ -24,18 +24,15 @@
 // #[macro_use]
 extern crate alloc;
 
-// Use `wee_alloc` as the global allocator.
-#[global_allocator]
-static ALLOC: wee_alloc::WeeAlloc = wee_alloc::WeeAlloc::INIT;
-
```
Comment (Robbepop) [inline core/src/lib.rs]: I assume `node-runtime` is using the same allocator?

File: examples/lang/runtime/.cargo/config [hunk 8]
```diff
@@ -0,0 +1,5 @@
+[target.wasm32-unknown-unknown]
+rustflags = [
+	"-C", "overflow-checks=on",
+	"-C", "link-args=-z stack-size=65536 --import-memory"
+]
```
Comment (Robbepop) [inline examples/lang/runtime/.cargo/config]: why this change?

File: examples/lang/runtime/Cargo.toml [hunk 9]
```diff
@@ -0,0 +1,31 @@
+[package]
+name = "runtime"
```
Comment (Robbepop) [inline examples/lang/runtime/Cargo.toml]: I would maybe call this example `calling_the_runtime` or something like that or `calling_back`

File: examples/lang/runtime/Cargo.toml [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+version = "0.1.0"
+authors = ["admin@parity.io"]
+edition = "2018"
+
+[dependencies]
+pdsl_core = { path = "../../../core" }
+pdsl_model = { path = "../../../model" }
+pdsl_lang = { path = "../../../lang" }
+parity-codec = { version = "3.2", default-features = false, features = ["derive"] }
```
Comment (Robbepop) [inline examples/lang/runtime/Cargo.toml]: I think above you used `parity-codec 3.3`

File: model/src/exec_env.rs [hunk 11]
```diff
@@ -125,6 +126,10 @@ impl EnvHandler {
         env::caller()
     }
 
+    pub fn dispatch_call(&self, call: Call) {
```
Comment (Robbepop) [inline model/src/exec_env.rs]: Missing documentation

File: core/Cargo.toml (hunk 1 above)
Comment (Robbepop) [reply core/Cargo.toml]: We can remove `wee_alloc` dependency here.
Also `pdsl_core` should not depend on `syn`.

**Comment by ascjones** [comment]
Closing this and splitting into 2 PRs. First of which is #72
